### PR TITLE
Make finalize step of recovery source non-blocking

### PR DESCRIPTION
**Pull request by dnhatn**
Relates #37291

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/ActionListenerResponseHandler.java [hunk 1]
```diff
@@ -37,10 +37,16 @@
...
 
-    public ActionListenerResponseHandler(ActionListener<? super Response> listener, Writeable.Reader<Response> reader) {
+    public ActionListenerResponseHandler(ActionListener<? super Response> listener, Writeable.Reader<Response> reader, String executor) {
         this.listener = Objects.requireNonNull(listener);
         this.reader = Objects.requireNonNull(reader);
+        this.executor = Objects.requireNonNull(executor);
+    }
+
+    public ActionListenerResponseHandler(ActionListener<? super Response> listener, Writeable.Reader<Response> reader) {
+        this(listener, reader,ThreadPool.Names.SAME);
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/action/ActionListenerResponseHandler.java]: extra space after `reader,` please

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 2]
```diff
@@ -232,16 +238,19 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
...
-                new RecoveryResponse(sendFileResult.phase1FileNames, sendFileResult.phase1FileSizes,
-                    sendFileResult.phase1ExistingFileNames, sendFileResult.phase1ExistingFileSizes, sendFileResult.totalSize,
-                    sendFileResult.existingTotalSize, sendFileResult.took.millis(), phase1ThrottlingWaitTime, prepareEngineTime.millis(),
-                    sendSnapshotResult.totalOperations, sendSnapshotResult.tookTime.millis())
-            );
+            final StepListener<Void> finalizeStep = new StepListener<>();
+            finalizeRecovery(sendSnapshotResult.targetLocalCheckpoint, finalizeStep);
+            finalizeStep.whenComplete(r -> {
+                assert resources.isEmpty() : "not every resource is released [" + resources + "]";
+                IOUtils.close(resources);
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: I would execute the `IOUtils.close(resources);` in a finally block after we sent back the response or the other way around.

**Review by s1monw — APPROVED** [review]
left some suggestions but LGTM otherwise. Much simpler to review now, thanks for the extra interation

File: server/src/main/java/org/elasticsearch/action/ActionListenerResponseHandler.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/ActionListenerResponseHandler.java]: I pushed https://github.com/elastic/elasticsearch/pull/37388/commits/44f796b00f6a9013a20374f294e0f8c06e262524.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java (hunk 2 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: I pushed https://github.com/elastic/elasticsearch/pull/37388/commits/ed215a9359d798fcd92cbdcf72a75f028f7cffeb.

**Comment by dnhatn** [comment]
Thanks @s1monw.

**Pull request merged**